File: cli/src/safeyolo/core/trace.py
"""Opt-in per-request execution tracing for the SafeYolo addon pipeline.

Answers the diagnostic question `/explain` cannot: which parts of the
pipeline actually executed for a given request? See issue #213 for the
motivation — missing audit events are ambiguous, missing trace steps
are explicit (`state=not_loaded` at read time).

Contract with the rest of the codebase:

- Callers use `record_step(flow, addon=..., hook=..., state=..., ...)`.
- The call is a no-op unless `flow.metadata.get("trace") is True`, so
  every request path pays only a single dict lookup when tracing is off.
- Any exception raised inside recording is swallowed here; addon
  enforcement is never disturbed by a trace-instrumentation bug.
- Nothing here writes to disk. Store is in-memory, bounded, TTL'd.

Reads happen via `TraceStore.get(request_id, agent_id)` and must be
scoped to the originating agent. The store deliberately returns None
both for unknown ids and for wrong-agent lookups, so an outside caller
cannot distinguish the two.
"""

from __future__ import annotations

import functools
import json
import logging
import os
import threading
import time
from collections import OrderedDict
from collections.abc import Callable
from dataclasses import asdict, dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mitmproxy import http

_log = logging.getLogger("safeyolo.trace")


# =============================================================================
# Configuration (env-tunable)
# =============================================================================

def _int_env(name: str, default: int) -> int:
    try:
        return int(os.environ.get(name, str(default)))
    except ValueError:
        return default


TRACE_TTL_S = _int_env("SAFEYOLO_TRACE_TTL_S", 300)
TRACE_GLOBAL_MAX = _int_env("SAFEYOLO_TRACE_GLOBAL_MAX", 1000)
TRACE_PER_AGENT_MAX = _int_env("SAFEYOLO_TRACE_PER_AGENT_MAX", 200)
TRACE_STEPS_MAX = _int_env("SAFEYOLO_TRACE_STEPS_MAX", 128)
TRACE_DETAILS_MAX_BYTES = _int_env("SAFEYOLO_TRACE_DETAILS_MAX_BYTES", 4096)


# =============================================================================
# Expected-addon registry
# =============================================================================
# The addons a normal outbound request should traverse. `TraceStore.get`
# emits `state=not_loaded` for any expected addon that never appeared in
# the recorded steps. Modules add themselves at import time by calling
# `register_expected_addon`; this keeps the "what should run" list beside
# the code that would run it rather than in a distant list.

_expected_addons: list[str] = []
_expected_lock = threading.Lock()


def register_expected_addon(name: str) -> None:
    """Declare an addon whose absence from a trace is diagnostic evidence.

    Safe to call more than once with the same name; duplicates are ignored.
    Ordering is insertion order (matches the mitmproxy addon load order in
    practice, which is what we compare against).
    """
    with _expected_lock:
        if name not in _expected_addons:
            _expected_addons.append(name)


def expected_addons() -> list[str]:
    with _expected_lock:
        return list(_expected_addons)


# =============================================================================
# Data model
# =============================================================================

# States a step can be in. Values chosen to match the vocabulary in issue #213
# so agent-side reasoning can enum-match rather than string-compare on drift.
STATE_EVALUATED = "evaluated"
STATE_BYPASSED = "bypassed"
STATE_ERROR = "error"
STATE_NOT_LOADED = "not_loaded"  # synthesised at read time, never recorded

# Canonical bypass/error reasons. Kept centrally so the skill DAG can branch on
# a stable, code-enforced vocabulary rather than free-form strings. Per-addon
# outcome constants live at the top of each addon module.
REASON_PRIOR_RESPONSE = "prior_response"      # an earlier addon already responded
REASON_POLICY_DISABLED = "policy_disabled"    # PDP said the addon is off for scope
REASON_ADDON_DISABLED = "addon_disabled"      # mitmproxy option disables the addon globally


@dataclass
class Step:
    """One instrumented step in the pipeline for a single request.

    Fields deliberately narrow — trace never carries request/response
    bodies, credential values, or authorization headers. `details` is
    for small non-secret categorical facts (match counts, decision
    strings) and is size-capped at serialise time.
    """
    addon: str
    hook: str  # "request" | "response"
    state: str  # STATE_EVALUATED | STATE_BYPASSED | STATE_ERROR
    outcome: str | None = None
    reason: str | None = None
    duration_us: int | None = None
    details: dict[str, Any] | None = None
    ts: float = field(default_factory=time.time)


@dataclass
class TraceRecord:
    request_id: str
    agent_id: str | None
    created_at: float = field(default_factory=time.time)
    steps: list[Step] = field(default_factory=list)
    truncated: bool = False  # step count exceeded TRACE_STEPS_MAX


# =============================================================================
# Store
# =============================================================================

class TraceStore:
    """In-memory, bounded, TTL'd store of per-request execution traces.

    Not persisted anywhere: this is short-lived diagnostic state whose
    only consumer is the originating agent (via /trace) and doctor.
    """

    def __init__(
        self,
        ttl_s: float = TRACE_TTL_S,
        global_max: int = TRACE_GLOBAL_MAX,
        per_agent_max: int = TRACE_PER_AGENT_MAX,
        steps_max: int = TRACE_STEPS_MAX,
        details_max_bytes: int = TRACE_DETAILS_MAX_BYTES,
    ) -> None:
        self._ttl_s = ttl_s
        self._global_max = global_max
        self._per_agent_max = per_agent_max
        self._steps_max = steps_max
        self._details_max_bytes = details_max_bytes
        # OrderedDict gives us O(1) LRU: move_to_end on write, popitem(last=False)
        # on eviction. Insertion order is refreshed on every append so the LRU
        # answers "least-recently-appended-to" — the natural notion here.
        self._records: OrderedDict[str, TraceRecord] = OrderedDict()
        self._by_agent: dict[str, list[str]] = {}
        self._lock = threading.Lock()

    # -- write side --------------------------------------------------------

    def _ensure_record(self, request_id: str, agent_id: str | None) -> TraceRecord:
        record = self._records.get(request_id)
        if record is None:
            record = TraceRecord(request_id=request_id, agent_id=agent_id)
            self._records[request_id] = record
            if agent_id:
                self._by_agent.setdefault(agent_id, []).append(request_id)
            self._enforce_caps(agent_id)
        else:
            # First writer wins on agent_id; later writers cannot claim a
            # trace they don't own even by racing. If agent_id was unknown
            # at creation, later fill it in (service-discovery may resolve
            # after the first step).
            if record.agent_id is None and agent_id:
                record.agent_id = agent_id
                self._by_agent.setdefault(agent_id, []).append(request_id)
            self._records.move_to_end(request_id)
        return record

    def append_step(
        self,
        request_id: str,
        agent_id: str | None,
        step: Step,
    ) -> None:
        with self._lock:
            self._expire_locked()
            record = self._ensure_record(request_id, agent_id)
            if len(record.steps) >= self._steps_max:
                record.truncated = True
                return
            record.steps.append(step)

    # -- read side ---------------------------------------------------------

    def get(self, request_id: str, agent_id: str | None) -> TraceRecord | None:
        """Return a trace only when `agent_id` owns it.

        Missing record and wrong-agent record return the same None so
        callers cannot use existence-of-id as an oracle.
        """
        with self._lock:
            self._expire_locked()
            record = self._records.get(request_id)
            if record is None:
                return None
            # None agent_id is not a wildcard: unresolved caller identity
            # must not read anyone's trace.
            if not agent_id or record.agent_id != agent_id:
                return None
            return record

    # -- housekeeping -----------------------------------------------------

    def _expire_locked(self) -> None:
        now = time.time()
        cutoff = now - self._ttl_s
        # OrderedDict is insertion-ordered; oldest is at the front. Records
        # get move_to_end on every append, so we can stop at the first live
        # one only for lookups by created_at. Because we care about
        # last-activity, not created_at, iterate from the front and drop
        # while their newest-step time (== record's move_to_end position)
        # is stale. Simpler: check both created_at and any step's ts.
        stale: list[str] = []
        for rid, rec in self._records.items():
            last = rec.steps[-1].ts if rec.steps else rec.created_at
            if last < cutoff:
                stale.append(rid)
            else:
                # Nothing older follows: rely on move_to_end keeping
                # active records toward the back.
                break
        for rid in stale:
            self._drop_locked(rid)

    def _enforce_caps(self, agent_id: str | None) -> None:
        # Per-agent cap first — a single agent can't monopolise the store.
        if agent_id:
            agent_ids = self._by_agent.get(agent_id, [])
            while len(agent_ids) > self._per_agent_max:
                oldest = agent_ids[0]
                self._drop_locked(oldest)
                # _drop_locked mutates agent_ids too; refresh reference in
                # case the list object was replaced by cleanup.
                agent_ids = self._by_agent.get(agent_id, [])
        # Global cap: LRU eviction from the front of the OrderedDict.
        while len(self._records) > self._global_max:
            oldest_rid, _ = next(iter(self._records.items()))
            self._drop_locked(oldest_rid)

    def _drop_locked(self, request_id: str) -> None:
        rec = self._records.pop(request_id, None)
        if rec is None:
            return
        if rec.agent_id:
            agent_ids = self._by_agent.get(rec.agent_id)
            if agent_ids:
                try:
                    agent_ids.remove(request_id)
                except ValueError:
                    pass
                if not agent_ids:
                    self._by_agent.pop(rec.agent_id, None)

    # -- serialisation helper (used by /trace) ----------------------------

    def serialise(self, record: TraceRecord) -> dict[str, Any]:
        """Build the wire payload for `/trace`, including `not_loaded` diff."""
        observed = {step.addon for step in record.steps}
        not_loaded = [name for name in expected_addons() if name not in observed]
        return {
            "request_id": record.request_id,
            "agent_id": record.agent_id,
            "created_at": record.created_at,
            "truncated": record.truncated,
            "steps": [self._safe_step_dict(step) for step in record.steps],
            "not_loaded": [
                {"addon": name, "state": STATE_NOT_LOADED}
                for name in not_loaded
            ],
        }

    def _safe_step_dict(self, step: Step) -> dict[str, Any]:
        payload = asdict(step)
        payload.pop("ts", None)
        details = payload.get("details")
        if details is not None:
            payload["details"] = self._cap_details(details)
        # Drop keys whose value is None to keep the wire payload compact
        # without hiding the semantic distinction between "no reason" and
        # "reason: ..." — both are keys the caller cares about only when
        # present.
        return {k: v for k, v in payload.items() if v is not None}

    def _cap_details(self, details: dict[str, Any]) -> dict[str, Any]:
        """Enforce the per-step details byte cap and drop nested structures."""
        cleaned: dict[str, Any] = {}
        for key, value in details.items():
            if isinstance(value, (str, int, bool)) or value is None:
                cleaned[key] = value
            else:
                # Anything else (dict, list, object) is coerced to a short str.
                # This prevents accidental inclusion of request/response
                # bodies or header dumps that a caller passed by reference.
                cleaned[key] = f"<{type(value).__name__}>"
        try:
            if len(json.dumps(cleaned)) > self._details_max_bytes:
                cleaned = {"_truncated": True}
        except (TypeError, ValueError):
            cleaned = {"_truncated": True}
        return cleaned


# =============================================================================
# Module-level singleton
# =============================================================================

_store: TraceStore | None = None
_store_lock = threading.Lock()


def get_store() -> TraceStore:
    global _store
    if _store is not None:
        return _store
    with _store_lock:
        if _store is None:
            _store = TraceStore()
    return _store


def reset_store_for_tests() -> None:
    """Drop the singleton so a test can construct a fresh one via `get_store`."""
    global _store
    with _store_lock:
        _store = None


# =============================================================================
# Convenience API (call-site facing)
# =============================================================================

def is_traced(flow: http.HTTPFlow) -> bool:
    try:
        return bool(flow.metadata.get("trace"))
    except Exception:  # pragma: no cover - defensive
        return False


def record_step(
    flow: http.HTTPFlow,
    *,
    addon: str,
    hook: str,
    state: str,
    outcome: str | None = None,
    reason: str | None = None,
    duration_us: int | None = None,
    details: dict[str, Any] | None = None,
) -> None:
    """Record one pipeline step. No-op unless the flow opted into tracing.

    Never raises. Trace-instrumentation failure must not weaken enforcement.
    """
    try:
        if not is_traced(flow):
            return
        request_id = flow.metadata.get("request_id")
        if not request_id:
            return
        agent_id = flow.metadata.get("agent")
        step = Step(
            addon=addon,
            hook=hook,
            state=state,
            outcome=outcome,
            reason=reason,
            duration_us=duration_us,
            details=details,
        )
        get_store().append_step(request_id, agent_id, step)
    except Exception as exc:  # noqa: BLE001 — enforcement must never regress
        _log.warning("trace record_step failed: %s: %s", type(exc).__name__, exc)


# =============================================================================
# Hook decorator — strictly observational: timing + error capture, nothing else
# =============================================================================
# Explicit per-hook wrapper rather than __init_subclass__ magic: keeps this
# security-relevant instrumentation visible at each participating addon (issue
# #213 review). Compatible with any addon that exposes a `name` attribute; not
# tied to SecurityAddon so `service-gateway` can use it too.
#
# Invariant (#213 review, second pass): a traced request MUST execute the
# exact same addon logic as an untraced one. The decorator does not decide
# whether the addon body runs, does not preempt short-circuits, and does not
# reinterpret decisions. It may:
#   - start a monotonic timer whose delta feeds `duration_us`,
#   - observe raised exceptions, record `state=error, reason=<ExcType>` and
#     re-raise so enforcement sees identical exception behaviour.
# It may NOT:
#   - skip calling the wrapped function under any condition,
#   - mutate flow.request / flow.response / flow.metadata beyond the private
#     timer key it sets and removes,
#   - emit `bypassed/*` on the addon's behalf (the addon owns its own
#     short-circuit decisions and their trace evidence).

def _hook_start_key(addon_name: str, hook: str) -> str:
    return f"_trace_hook_start:{addon_name}:{hook}"


def _elapsed_us_from(flow: http.HTTPFlow, addon_name: str, hook: str) -> int | None:
    """Compute microseconds elapsed since `trace_addon_hook` started this hook.

    Returns None if the timer wasn't started (untraced flow, or the addon
    wasn't decorated) so callers can distinguish "measured" from "unavailable"
    in the trace payload. Defensive against Mock/duck-typed flow.metadata:
    a non-int `start` reads as "no timer" rather than raising.
    """
    try:
        start = flow.metadata.get(_hook_start_key(addon_name, hook))
    except Exception:  # pragma: no cover - defensive
        return None
    if not isinstance(start, int):
        return None
    return (time.perf_counter_ns() - start) // 1000


def trace_addon_hook(hook: str) -> Callable:
    """Observational decorator for an addon's `request` or `response` method.

    See the invariant block above. This wrapper only:
      1. starts a monotonic timer that the addon's own `_trace_evaluated` /
         `_trace_bypassed` calls read to fill `duration_us`, and
      2. catches exceptions to emit `state=error, reason=<ExceptionType>,
         duration_us=<measured>` before re-raising so enforcement sees the
         identical exception.

    The wrapped function is ALWAYS called for a traced flow — the decorator
    does not decide whether the hook body runs. Addons that short-circuit
    (e.g. `if flow.response: return` or `if self.is_bypassed(flow): return`)
    are responsible for emitting their own `bypassed/*` trace evidence.
    """
    if hook not in ("request", "response"):
        raise ValueError(f"trace_addon_hook: unsupported hook {hook!r}")

    def decorator(fn: Callable) -> Callable:
        @functools.wraps(fn)
        def wrapper(self, flow, *args, **kwargs):
            if not is_traced(flow):
                return fn(self, flow, *args, **kwargs)

            addon_name = getattr(self, "name", type(self).__name__)
            key = _hook_start_key(addon_name, hook)
            flow.metadata[key] = time.perf_counter_ns()
            try:
                return fn(self, flow, *args, **kwargs)
            except BaseException as exc:
                # Compute duration BEFORE record_step so the trace step reflects
                # how long the failed hook ran.
                elapsed = _elapsed_us_from(flow, addon_name, hook)
                try:
                    record_step(
                        flow,
                        addon=addon_name,
                        hook=hook,
                        state=STATE_ERROR,
                        reason=type(exc).__name__,
                        duration_us=elapsed,
                    )
                except Exception as inner:  # noqa: BLE001
                    _log.warning(
                        "trace_addon_hook error emit failed: %s: %s",
                        type(inner).__name__, inner,
                    )
                raise
            finally:
                # Clean up the private timer key so it doesn't leak into other
                # hooks or into serialisation surfaces that inspect
                # flow.metadata. The delete is the only mutation this wrapper
                # makes to flow.metadata visible to production code.
                flow.metadata.pop(key, None)
        return wrapper
    return decorator


# =============================================================================
# Module-level convenience for addons that aren't SecurityAddon subclasses
# =============================================================================
# service-gateway needs to record trace without inheriting base.py plumbing.

def trace_evaluated(
    flow: http.HTTPFlow,
    *,
    addon: str,
    hook: str = "request",
    outcome: str,
    duration_us: int | None = None,
    **details: Any,
) -> None:
    """Standalone `state=evaluated` emitter for non-SecurityAddon participants."""
    if duration_us is None:
        duration_us = _elapsed_us_from(flow, addon, hook)
    record_step(
        flow,
        addon=addon,
        hook=hook,
        state=STATE_EVALUATED,
        outcome=outcome,
        duration_us=duration_us,
        details=details or None,
    )


def trace_bypassed(
    flow: http.HTTPFlow,
    *,
    addon: str,
    hook: str = "request",
    reason: str,
) -> None:
    """Standalone `state=bypassed` emitter for non-SecurityAddon participants."""
    record_step(
        flow,
        addon=addon,
        hook=hook,
        state=STATE_BYPASSED,
        reason=reason,
    )


# No mitmproxy addon here — this module is pure infrastructure, imported
# by base.py and agent_api.py. Explicit empty list keeps `-s trace.py`
# from doing anything surprising.
addons: list = []

File: proxy/src/trace.rs
//! In-memory source trace records. The caller owns opt-in, trusted identity,
//! clock sampling and runtime hook reachability; this store schedules no work.

use std::{fmt, sync::Mutex};

use indexmap::IndexMap;
use num_bigint::BigInt;
use serde_json::{Value, json};
use zeroize::Zeroize;

use crate::circuits::{CircuitValue as C, ErrorKind as NumericError};

const EXPECTED_ADDONS: [&str; 6] = [
    "service-gateway",
    "network-guard",
    "circuit-breaker",
    "credential-guard",
    "pattern-scanner",
    "test-context",
];

#[derive(Clone)]
pub struct Settings {
    pub ttl_s: C,
    pub global_max: BigInt,
    pub per_agent_max: BigInt,
    pub steps_max: BigInt,
    pub details_max_bytes: BigInt,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            ttl_s: 300.into(),
            global_max: 1000.into(),
            per_agent_max: 200.into(),
            steps_max: 128.into(),
            details_max_bytes: 4096.into(),
        }
    }
}

impl Settings {
    /// The runtime samples these once when it creates the shared store.
    /// Invalid integers, including non-Unicode environment bytes, use defaults.
    pub fn from_env() -> Self {
        let read = |name| std::env::var(name).ok();
        let ttl = read("SAFEYOLO_TRACE_TTL_S");
        let global = read("SAFEYOLO_TRACE_GLOBAL_MAX");
        let agent = read("SAFEYOLO_TRACE_PER_AGENT_MAX");
        let steps = read("SAFEYOLO_TRACE_STEPS_MAX");
        let details = read("SAFEYOLO_TRACE_DETAILS_MAX_BYTES");
        Self::from_environment_values(
            ttl.as_deref(),
            global.as_deref(),
            agent.as_deref(),
            steps.as_deref(),
            details.as_deref(),
        )
    }

    pub fn from_environment_values(
        ttl_s: Option<&str>,
        global_max: Option<&str>,
        per_agent_max: Option<&str>,
        steps_max: Option<&str>,
        details_max_bytes: Option<&str>,
    ) -> Self {
        let integer = |value: Option<&str>, default: i32| {
            value
                .and_then(|value| {
                    crate::flow_store::integer(&C::Other(Value::String(value.into())))
                })
                .unwrap_or_else(|| default.into())
        };
        Self {
            ttl_s: integer(ttl_s, 300).into(),
            global_max: integer(global_max, 1000),
            per_agent_max: integer(per_agent_max, 200),
            steps_max: integer(steps_max, 128),
            details_max_bytes: integer(details_max_bytes, 4096),
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ErrorKind {
    Poisoned,
    Type,
    Attribute,
    Overflow,
    Compatibility,
    Index,
    StopIteration,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Error(ErrorKind);
impl Error {
    pub fn kind(self) -> ErrorKind {
        self.0
    }
}
impl fmt::Display for Error {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("trace store operation failed")
    }
}
impl std::error::Error for Error {}
pub type Result<T> = std::result::Result<T, Error>;

/// Concrete source Step fields. Details accept ordered JSON-shaped values;
/// strings/integers/bools/null survive, while floats and containers are named.
pub struct Step {
    pub addon: String,
    pub hook: String,
    pub state: String,
    pub outcome: Option<String>,
    pub reason: Option<String>,
    pub duration_us: Option<BigInt>,
    pub details: Option<C>,
    pub ts: f64,
    pub connection_id: Option<String>,
    pub method: Option<String>,
    pub host: Option<String>,
    pub port: Option<BigInt>,
}

impl Step {
    pub fn new(
        addon: impl Into<String>,
        hook: impl Into<String>,
        state: impl Into<String>,
        ts: f64,
    ) -> Self {
        Self {
            addon: addon.into(),
            hook: hook.into(),
            state: state.into(),
            outcome: None,
            reason: None,
            duration_us: None,
            details: None,
            ts,
            connection_id: None,
            method: None,
            host: None,
            port: None,
        }
    }

    fn document(&self, limit: &BigInt) -> Result<Value> {
        let mut fields = serde_json::Map::new();
        fields.insert("addon".into(), json!(self.addon));
        fields.insert("hook".into(), json!(self.hook));
        fields.insert("state".into(), json!(self.state));
        optional(&mut fields, "outcome", self.outcome.as_deref());
        optional(&mut fields, "reason", self.reason.as_deref());
        if let Some(value) = &self.duration_us {
            fields.insert("duration_us".into(), integer_json(value));
        }
        if let Some(value) = &self.details
            && !matches!(value, C::Other(Value::Null))
        {
            fields.insert("details".into(), cap_details(value, limit)?);
        }
        optional(&mut fields, "connection_id", self.connection_id.as_deref());
        optional(&mut fields, "method", self.method.as_deref());
        optional(&mut fields, "host", self.host.as_deref());
        if let Some(value) = &self.port {
            fields.insert("port".into(), integer_json(value));
        }
        Ok(Value::Object(fields))
    }
}

impl Drop for Step {
    fn drop(&mut self) {
        self.addon.zeroize();
        self.hook.zeroize();
        self.state.zeroize();
        for value in [
            &mut self.outcome,
            &mut self.reason,
            &mut self.connection_id,
            &mut self.method,
            &mut self.host,
        ]
        .into_iter()
        .flatten()
        {
            value.zeroize();
        }
    }
}

struct Record {
    agent_id: Option<String>,
    created_at: f64,
    steps: Vec<Step>,
    truncated: bool,
}
impl Drop for Record {
    fn drop(&mut self) {
        if let Some(agent) = &mut self.agent_id {
            agent.zeroize();
        }
    }
}

#[derive(Default)]
struct State {
    records: IndexMap<String, Record>,
    by_agent: IndexMap<String, Vec<String>>,
}

impl State {
    fn drop_record(&mut self, request_id: &str) {
        let Some((mut id, record)) = self.records.shift_remove_entry(request_id) else {
            return;
        };
        id.zeroize();
        if let Some(agent) = record.agent_id.as_deref().filter(|agent| !agent.is_empty()) {
            let empty = if let Some(ids) = self.by_agent.get_mut(agent) {
                if let Some(index) = ids.iter().position(|id| id == request_id) {
                    ids.remove(index).zeroize();
                }
                ids.is_empty()
            } else {
                false
            };
            if empty && let Some((mut name, _)) = self.by_agent.shift_remove_entry(agent) {
                name.zeroize();
            }
        }
    }

    fn expire(&mut self, settings: &Settings, now: f64) -> Result<()> {
        // Division by float one reuses the existing Python numeric conversion,
        // including Overflow for an integer too large to convert to float.
        let C::Float(ttl) = settings
            .ttl_s
            .divide(&C::Float(1.0))
            .map_err(numeric_error)?
        else {
            unreachable!("floating divisor")
        };
        let cutoff = now - ttl;
        let count = self
            .records
            .values()
            .take_while(|record| {
                record
                    .steps
                    .last()
                    .map_or(record.created_at, |step| step.ts)
                    < cutoff
            })
            .count();
        // Source deliberately stops at the first live record. Append order and
        // retained step timestamps can disagree; do not scan beyond that point.
        for _ in 0..count {
            let id = self.records.first().expect("counted record").0.clone();
            self.drop_record(&id);
        }
        Ok(())
    }

    fn enforce_caps(&mut self, settings: &Settings, agent: Option<&str>) -> Result<()> {
        if let Some(agent) = agent.filter(|agent| !agent.is_empty()) {
            loop {
                let ids = self.by_agent.get(agent);
                if BigInt::from(ids.map_or(0, Vec::len)) <= settings.per_agent_max {
                    break;
                }
                let oldest = ids
                    .and_then(|ids| ids.first())
                    .ok_or(Error(ErrorKind::Index))?
                    .clone();
                self.drop_record(&oldest);
            }
        }
        while BigInt::from(self.records.len()) > settings.global_max {
            let oldest = self
                .records
                .first()
                .ok_or(Error(ErrorKind::StopIteration))?
                .0
                .clone();
            self.drop_record(&oldest);
        }
        Ok(())
    }
}

impl Drop for State {
    fn drop(&mut self) {
        for (mut id, _) in self.records.drain(..) {
            id.zeroize();
        }
        for (mut agent, mut ids) in self.by_agent.drain(..) {
            agent.zeroize();
            for id in &mut ids {
                id.zeroize();
            }
        }
    }
}

pub struct TraceStore {
    settings: Settings,
    state: Mutex<State>,
}

impl TraceStore {
    pub fn new(settings: Settings) -> Self {
        Self {
            settings,
            state: Mutex::new(State::default()),
        }
    }

    pub fn append(
        &self,
        request_id: &str,
        agent: Option<&str>,
        step: Step,
        now: f64,
    ) -> Result<()> {
        let mut state = self.state.lock().map_err(|_| Error(ErrorKind::Poisoned))?;
        state.expire(&self.settings, now)?;
        if let Some((id, mut record)) = state.records.shift_remove_entry(request_id) {
            // Only None can be filled; an empty first owner is not None.
            if record.agent_id.is_none()
                && let Some(agent) = agent.filter(|agent| !agent.is_empty())
            {
                record.agent_id = Some(agent.into());
                state
                    .by_agent
                    .entry(agent.into())
                    .or_default()
                    .push(request_id.into());
            }
            // Existing records move even when the new step will be capped.
            // Late owner fill does not enforce caps in the source.
            state.records.insert(id, record);
        } else {
            state.records.insert(
                request_id.into(),
                Record {
                    agent_id: agent.map(str::to_owned),
                    created_at: now,
                    steps: Vec::new(),
                    truncated: false,
                },
            );
            if let Some(agent) = agent.filter(|agent| !agent.is_empty()) {
                state
                    .by_agent
                    .entry(agent.into())
                    .or_default()
                    .push(request_id.into());
            }
            state.enforce_caps(&self.settings, agent)?;
        }
        // A zero cap may already have removed the just-created source record.
        // Its subsequent orphan append has no observable store effect.
        if let Some(record) = state.records.get_mut(request_id) {
            if BigInt::from(record.steps.len()) >= self.settings.steps_max {
                record.truncated = true;
            } else {
                record.steps.push(step);
            }
        }
        Ok(())
    }

    /// Missing, unresolved and foreign-owner reads all return None after expiry.
    /// The returned report is an owned snapshot; reads do not refresh LRU order.
    pub fn get(&self, request_id: &str, agent: Option<&str>, now: f64) -> Result<Option<Value>> {
        let mut state = self.state.lock().map_err(|_| Error(ErrorKind::Poisoned))?;
        state.expire(&self.settings, now)?;
        let Some(record) = state.records.get(request_id) else {
            return Ok(None);
        };
        if agent.is_none_or(str::is_empty) || record.agent_id.as_deref() != agent {
            return Ok(None);
        }
        let created_at = serde_json::Number::from_f64(record.created_at)
            .ok_or(Error(ErrorKind::Compatibility))?;
        let expected: &[&str] = if record.steps.iter().any(|step| step.hook == "http_connect") {
            &["network-guard"]
        } else {
            &EXPECTED_ADDONS
        };
        let not_loaded: Vec<_> = expected
            .iter()
            .filter(|name| !record.steps.iter().any(|step| &step.addon == *name))
            .map(|name| json!({"addon":name,"state":"not_loaded"}))
            .collect();
        let steps = record
            .steps
            .iter()
            .map(|step| step.document(&self.settings.details_max_bytes))
            .collect::<Result<Vec<_>>>()?;
        Ok(Some(
            json!({"request_id":request_id,"agent_id":record.agent_id,
            "created_at":created_at,"truncated":record.truncated,"steps":steps,"not_loaded":not_loaded}),
        ))
    }
}

fn numeric_error(error: crate::circuits::Error) -> Error {
    Error(match error.kind() {
        NumericError::Type => ErrorKind::Type,
        NumericError::Overflow => ErrorKind::Overflow,
        _ => ErrorKind::Compatibility,
    })
}
fn optional(fields: &mut serde_json::Map<String, Value>, name: &str, value: Option<&str>) {
    if let Some(value) = value {
        fields.insert(name.into(), json!(value));
    }
}
fn integer_json(value: &BigInt) -> Value {
    Value::Number(value.to_string().parse().expect("BigInt JSON integer"))
}

fn cap_details(details: &C, limit: &BigInt) -> Result<Value> {
    if let C::Other(Value::Object(_)) = details {
        return cap_details(&C::from(details.json().map_err(numeric_error)?), limit);
    }
    let fields = details.as_object().ok_or(Error(ErrorKind::Attribute))?;
    let mut cleaned = serde_json::Map::new();
    for (key, value) in fields {
        let value = match value {
            C::Integer(value) => {
                if value.to_string().trim_start_matches('-').len() > 4300 {
                    return Ok(json!({"_truncated":true}));
                }
                integer_json(value)
            }
            C::Bool(value) => json!(value),
            C::Float(_) => json!("<float>"),
            C::Array(_) => json!("<list>"),
            C::Object(_) => json!("<dict>"),
            C::Other(Value::String(value)) => json!(value),
            C::Other(Value::Null) => Value::Null,
            C::Other(Value::Bool(value)) => json!(value),
            C::Other(Value::Array(_)) => json!("<list>"),
            C::Other(Value::Object(_)) => json!("<dict>"),
            C::Other(Value::Number(value)) => {
                let text = value.to_string();
                if text.contains(['.', 'e', 'E']) {
                    json!("<float>")
                } else if text.trim_start_matches('-').len() > 4300 {
                    return Ok(json!({"_truncated":true}));
                } else {
                    Value::Number(value.clone())
                }
            }
            C::Temporal(_) => return Err(Error(ErrorKind::Compatibility)),
        };
        cleaned.insert(key.clone(), value);
    }
    let cleaned = Value::Object(cleaned);
    // Source names this a byte cap, but counts default json.dumps characters.
    // The shared formatter emits ASCII escapes, so its byte length is exact.
    if BigInt::from(crate::python_json::encoded_len(&cleaned)) > *limit {
        Ok(json!({"_truncated":true}))
    } else {
        Ok(cleaned)
    }
}

#[cfg(test)]
mod tests;
